src/kubernetes/reflector.rs

```diff
@@ -0,0 +1,569 @@
+            label_selector: self.label_selector.as_ref().map(|s| s.as_str()),
+            pretty: None,
+            resource_version: self.resource_version.get(),
+            timeout_seconds: None,
+            allow_watch_bookmarks: Some(true),
+        };
+        let stream = self.watcher.watch(watch_optional).await?;
+        Ok(stream)
+    }
+
+    /// Translate received watch event to the state update.
+    fn process_event(&mut self, event: WatchEvent<<W as Watcher>::Object>) {
+        match event {
+            WatchEvent::Added(object) => {
+                if let Some((key, value)) = kv(object) {
+                    trace!(message = "got an object event", uid = ?key, event = "added");
+                    self.state_writer.insert(key, value);
+                } else {
+                    warn!(
+                        message = "got an object event but unable to obtain a key from object",
+                        event = "added",
+                    );
+                }
+            }
+            WatchEvent::Deleted(object) => {
+                if let Some((key, _value)) = kv(object) {
+                    trace!(message = "got an object event", uid = ?key, event = "deleted");
+                    self.state_writer.empty(key);
+                } else {
+                    warn!(
+                        message = "got an object event but unable to obtain a key from object",
+                        event = "deleted",
+                    );
+                }
+            }
+            WatchEvent::Modified(object) => {
+                if let Some((key, value)) = kv(object) {
+                    trace!(message = "got an object event", uid = ?key, event = "modified");
+                    self.state_writer.update(key, value);
+                } else {
+                    warn!(
+                        message = "got an object event but unable to obtain a key from object",
+                        event = "modified",
+                    );
+                }
+            }
+            WatchEvent::Bookmark(_object) => {
+                // noop
+                trace!(message = "got an object event", event = "bookmark");
+            }
+            _ => unreachable!("other event types should never reach this code"),
+        }
+    }
+}
+
+/// Errors that can occur while watching.
+#[derive(Debug, Snafu)]
+pub enum Error<I, S>
+where
+    I: std::error::Error + 'static,
+    S: std::error::Error + 'static,
+{
+    /// Returned when the watch invocation (HTTP request) failed.
+    #[snafu(display("watch invocation failed"))]
+    Invocation {
+        /// The underlying invocation error.
+        source: I,
+    },
+
+    /// Returned when the stream failed with an error.
+    #[snafu(display("streaming error"))]
+    Streaming {
+        /// The underlying stream error.
+        source: S,
+    },
+}
+
+/// An alias to the value used at [`evmap`].
+pub type Value<T> = Box<HashValue<T>>;
```

Comment:
I could see confusion about this type being `vector::Value` in to readers in other code.

Reply:
Hmm, good point. The name `Value` here comes from the `evmap::Values`, which is, naturally, a collection of `Value`s. Given that `evmap::Values` is not under our control, it'd be impossible fully eliminate the confusion.

We can rename this `Value` to `EvmapValue` though. Would this help? What do you think?

Reply:
Yeah you're right :( Let's leave it.